### PR TITLE
Return null for aggregations factory even if empty aggs is present in search source

### LINKED ISSUE #8115: [BUG] Opensearch Dashboards getting 500 internal server error while fetching search aggregation data
**Describe the bug**
Referring to the below commit which is making version specific changes to 2.8.0. Looks like for Dashboards versions above 2.8.0,  this is a breaking change. Currently, we are seeing functional test failures which might be related to this on both windows and linux while running on main (3.0.0) version. We don't see this issue in our 2.x branch (which is currently not bumped to 2.8.1 yet). 

https://github.com/opensearch-project/OpenSearch/pull/7514

**To Reproduce**

We can reproduce the error in local with sample flight test data and load the Dashboard page. 

Screenshot 2023-06-16 at 11 20 08 AM

```
{
	"statusCode": 500,
	"error": "Internal Server Error",
	"message": "search_phase_execution_exception: [response_handler_failure_transport_exception] Reason: java.lang.NullPointerException: Cannot invoke \"org.opensearch.common.io.stream.DelayableWriteable.asSerialized(org.opensearch.common.io.stream.Writeable$Reader, org.opensearch.common.io.stream.NamedWriteableRegistry)\" because the return value of \"org.opensearch.search.query.QuerySearchResult.aggregations()\" is null",
	"attributes": {
		"error": {
			"root_cause": [{
				"type": "response_handler_failure_transport_exception",
				"reason": "java.lang.NullPointerException: Cannot invoke \"org.opensearch.common.io.stream.DelayableWriteable.asSerialized(org.opensearch.common.io.stream.Writeable$Reader, org.opensearch.common.io.stream.NamedWriteableRegistry)\" because the return value of \"org.opensearch.search.query.QuerySearchResult.aggregations()\" is null"
			}],
			"type": "search_phase_execution_exception",
			"reason": "",
			"phase": "fetch",
			"grouped": true,
			"failed_shards": [{
				"shard": 0,
				"index": "opensearch_dashboards_sample_data_flights",
				"node": "91-QImtGTB2vg2_m2Y07HA",
				"reason": {
					"type": "response_handler_failure_transport_exception",
					"reason": "java.lang.NullPointerException: Cannot invoke \"org.opensearch.common.io.stream.DelayableWriteable.asSerialized(org.opensearch.common.io.stream.Writeable$Reader, org.opensearch.common.io.stream.NamedWriteableRegistry)\" because the return value of \"org.opensearch.search.query.QuerySearchResult.aggregations()\" is null",
					"caused_by": {
						"type": "null_pointer_exception",
						"reason": "Cannot invoke \"org.opensearch.common.io.stream.DelayableWriteable.asSerialized(org.opensearch.common.io.stream.Writeable$Reader, org.opensearch.common.io.stream.NamedWriteableRegistry)\" because the return value of \"org.opensearch.search.query.QuerySearchResult.aggregations()\" is null"
					}
				}
			}],
			"caused_by": {
				"type": "null_pointer_exception",
				"reason": "Cannot invoke \"org.opensearch.search.aggregations.InternalAggregations.getSerializedSize()\" because \"reducePhase.aggregations\" is null"
			}
		}
	}
}
```

**Expected behavior**
Dashboards should be able to consume search pipeline API response without any issue. 

Opensearch Dashboards version   > 2.8.0

**Additional context**

### PR DESCRIPTION
**Pull request by jed326**
### Description
Whenever `"aggs"` is passed into the request body, even if it is empty (ex. `"aggs": {},`), then this line will evaluate to true:

https://github.com/opensearch-project/OpenSearch/blob/8eea7b986453271cd227a7b98ee09e70d5e28634/server/src/main/java/org/opensearch/action/search/QueryPhaseResultConsumer.java#L117

Causing us to go down the wrong path and hit the NPE.

This behavior exists from before #7514 and was handled by various checks for the aggregators, for example like here: https://github.com/opensearch-project/OpenSearch/blob/2.8/server/src/main/java/org/opensearch/search/aggregations/AggregationPhase.java#L68-L74 where collectors are not added if the aggregators length is not 0.

### Related Issues
Resolves #8115

### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/search/aggregations/AggregatorFactories.java [hunk 1]
```diff
@@ -235,7 +235,7 @@ private static AggregatorFactories.Builder parseAggregators(XContentParser parse
             }
         }
 
-        return factories;
+        return factories.count() > 0 ? factories : null;
```
Comment (jed326) [inline server/src/main/java/org/opensearch/search/aggregations/AggregatorFactories.java:238]: This value is checked for null to determine if there are any aggregations present or not.
https://github.com/opensearch-project/OpenSearch/blob/8eea7b986453271cd227a7b98ee09e70d5e28634/server/src/main/java/org/opensearch/action/search/QueryPhaseResultConsumer.java#L117

File: rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml [hunk 2]
```diff
@@ -528,3 +528,31 @@ setup:
   - is_false:  aggregations.unsigned_long_range.buckets.2.to
 
   - match: { aggregations.unsigned_long_range.buckets.2.doc_count: 0 }
+
+---
+"Empty Agg Body":
```
Comment (jed326) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml:533]: Verified that this test fails without the changes in this PR

**Comment by jed326** [comment]
@reta @andrross @sohami Could you please take a look? This addresses the bug identified in #8115.

**Review by sohami — APPROVED** [review]
lgtm

File: CHANGELOG.md [hunk 3]
```diff
@@ -69,6 +69,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Support OpenSSL Provider with default Netty allocator ([#5460](https://github.com/opensearch-project/OpenSearch/pull/5460))
 - Replaces ZipInputStream with ZipFile to fix Zip Slip vulnerability ([#7230](https://github.com/opensearch-project/OpenSearch/pull/7230))
 - Add missing validation/parsing of SearchBackpressureMode of SearchBackpressureSettings ([#7541](https://github.com/opensearch-project/OpenSearch/pull/7541))
+- Change Aggregator Factories to return null if empty aggs body is present in search request ([#8206](https://github.com/opensearch-project/OpenSearch/pull/8206))
```
Comment (andrross) [inline CHANGELOG.md:72]: Two things:

- This fixes a bug for an unreleased change, so we don't need a changelog entry as this bug was never visible in a release.
- Not relevant here, but for future reference, any change that is intended to be backported to 2.x should be in the `[Unreleased 2.x]` section of the changelog. See [CONTRIBUTING.md](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#changelog) for comprehensive guidance on the changelog.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml (hunk 2 above)
Comment (andrross) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml:533]: Thanks for adding a failing test!

Nitpick, but why include this case in the `range` test?

Comment (andrross) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml:533]: Also, can we add some sort of match to the test, like maybe on `hits.total`?

Comment (sohami) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml:533]: Actually this file is for `aggregation` test with ranges. Full path includes the `search.aggregation` folder: `rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml`

Comment (reta) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/40_range.yml:533]: @jed326 we could move it off to separate file like `60_empty.yml`

**Comment by jed326** [comment]
@andrross @reta Fixed the changelog and moved the test to a new file with a basic hits.total match. Let me know what you think!

**Comment by jed326** [comment]
Flakey test from [here](https://github.com/opensearch-project/OpenSearch/pull/8206#issuecomment-1601846811) is due to the change needing to be backported to 2.x as well.
From logs https://build.ci.opensearch.org/job/gradle-check/18133/consoleText we can see that this test was for 2.9 which we would expect to fail at this point.

**Comment by jed326** [comment]
The test failures seem unrelated to these changes. @reta @andrross do you think we could get this merged tonight to unblock OpenSearch Dashboards builds?

**Review by wbeckler — APPROVED** [review]
Nice way to add a test. Thank you for getting this in. I hope it gets approved ASAP.

**Review by Rishikesh1159 — APPROVED** [review]

File: rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/60_empty.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+---
+"Empty Agg Body":
```
Comment (reta) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/60_empty.yml:2]: Super nit, we return empty `aggs`:

```suggestion
"Empty aggs Body":
```

**Comment by jed326** [comment]
@reta Applied your suggestion, I think we're good to merge then? Thanks!

**Review by andrross — APPROVED** [review]

**Pull request merged**